addons/project/models/project_project.py

```diff
@@ -415,6 +415,16 @@ def create(self, vals_list):
         return projects
 
     def write(self, vals):
+        # Check if user can move a project in a stage owned by selected company
+        if vals.get('stage_id'):
+            new_stage_company = self.env["project.project.stage"].browse(vals.get('stage_id')).company_id
+
+            if new_stage_company and self.company_id != new_stage_company:
+                raise exceptions.UserError(
+                    _("This project is associated with %s, whereas the selected stage belongs to %s. "
+                    "There are a couple of options to consider: either remove the company designation from the project or from the stage. "
+                    "Alternatively, you can update the company information for these records to align them under the same company.", self.company_id.name, new_stage_company.name))
```

Comment:
Hmm, after checking again I don't think this should be here, because I don't think we should allow projects without company_id to be in a stage that has a company_id

Reply:
Alternatively, we could do a python constraint for that part. 🙂 

```py
@api.constrains("stage_id", "company_id")
def _ensure_stage_has_same_company(self):
    if not all(p.stage_id.company_id == p.company_id for p in self):
        raise UserError(
            _("This project is associated with %s, whereas the selected stage belongs to %s. "
            "There are a couple of options to consider: either remove the company designation from the project or from the stage. "
            "Alternatively, you can update the company information for these records to align them under the same company.", self.company_id.name, new_stage_company.name)
        )
```
🙂